en/02_Developer_Guides/03_Forms/07_Using_GridField_With_Arbitrary_Data.md

```diff
@@ -0,0 +1,437 @@
+
+[info]
+These will both use the field list you [passed into `GridFieldDataColumns`](#arraydata-display-as-rows) to know which fields they should use - though you can explicitly call [`GridFieldExportButton::setExportColumns()`](api:SilverStripe\Forms\GridField\GridFieldExportButton::setExportColumns()) and [`GridFieldPrintButton::setPrintColumns()`](api:SilverStripe\Forms\GridField\GridFieldPrintButton::setPrintColumns()) if you want to export/print different columns than those displayed in the grid view.
+[/info]
+
+```php
+use SilverStripe\Forms\GridField\GridFieldExportButton;
+use SilverStripe\Forms\GridField\GridFieldPrintButton;
+$gridField->getConfig()->addComponents([
+    GridFieldExportButton::create('buttons-before-left'),
+    GridFieldPrintButton::create('buttons-before-left'),
+]);
+```
+
+### Viewing data in a form {#arraydata-view}
+
+For data to be viewed in a read-only form, each record in the list must have an `ID` field, and the value of that field must be a positive integer. This is used in the URL for the form. Without it, the gridfield has no way to know which record it should be displaying in the form.
+
+You'll need to add a `GridFieldDetailForm` component to the `GridField` and tell it how to represent your data by passing a [`FieldList`](api:SilverStripe\Forms\FieldList) into [`GridFieldDetailForm::setFields()`](api:SilverStripe\Forms\GridField\GridFieldDetailForm::setFields()).
+
+[hint]
+Because `ArrayData` doesn't implement a `canEdit()` method, the form will be implicitly turned into a read-only form for you. You don't need to worry about passing in read-only form fields.
+[/hint]
+
+```php
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\GridField\GridFieldDetailForm;
+use SilverStripe\Forms\GridField\GridFieldViewButton;
+use SilverStripe\Forms\HiddenField;
+use SilverStripe\Forms\TextField;
+
+$detailForm = GridFieldDetailForm::create();
+$detailForm->setFields(FieldList::create([
+    HiddenField::create('ID'),
+    TextField::create('FeldName', 'View Field Label'),
+]));
+$gridField->getConfig()->addComponents([
+    GridFieldViewButton::create(),
+    $detailForm,
+]);
+```
+
+## Representing data in your own class
+
+As mentioned in the preamble above, the class representing your data must be a subclass of `ViewableData` in order for it to be used in a `GridField`. It's possible that some components will work even if your class isn't a subclass of `ViewableData`, but that use case is not supported and could break at any time.
+
+[info]
+Note that all of the methods that this documentation implements can be ommitted, with exception of the [editing data in a form](#custom-edit) section.
+
+However, if you omit these method implementations, you must instead pass the required information through to the relevant `GridField` components as shown in [representing data with `ArrayData`](#representing-data-with-arraydata) above.
+[/info]
```

Comment:
```suggestion
All of the methods describe in this arcticle can be ommitted, with exception of the [editing data in a form](#custom-edit) section.

However, if you omit these method implementations, you must pass the required information through to the relevant `GridField` components as shown in [representing data with `ArrayData`](#representing-data-with-arraydata) above.
```

Reply:
Done. But can you please explain when you think `info` callout blocks should be used? It seems like you have a specific set of criteria in your head that should be documented in the contributing docs section.

Reply:
The [description you added](https://docs.silverstripe.org/en/5/contributing/documentation/#callout-blocks) in the contribution doc:
> "Info box": An info box is useful for adding, deepening or accenting information in the main text. They can be used for background knowledge, or to provide links to further information (ie, a "see also" link).

But maybe we have different interpretation on it. In my mind the content in the infobox should make sense even if it's read out of context. Equally, the surrounding content to the info box should a make sense even if the info box is not there.

In this context, I don't think the info box content could stand alone. I also don't think you could read the _the class representing your data must be a subclass of `ViewableData`_ by itself and understand what's expected.

Reply:
What I wrote there was intended to be an example of the sorts of things that the callout blocks _might_ be useful for, not an exhaustive list of what it _is allowed_ to be used for.